[AMDGPU][MC] Disallow null as saddr in flat instructions

jwanggit86 opened the pull request:
Some flat instructions have an saddr operand. When 'null' is provided as saddr, it may have the same encoding as another instruction. For example, the instructions 'global_atomic_add v1, v2, null' and 'global_atomic_add v[1:2], v2, off' have the same encoding. This patch disallows having null as saddr.

jwanggit86 committed:
[AMDGPU][MC] Disallow null as saddr in flat instructions

Some flat instructions have an saddr operand. When 'null' is provided
as saddr, it may have the same encoding as another instruction. For
example, the instructions 'global_atomic_add v1, v2, null' and
'global_atomic_add v[1:2], v2, off' have the same encoding. This patch
disallows having null as saddr.

jwanggit86 commented:
For Issue [48508](https://github.com/llvm/llvm-project/issues/48508)

rovka commented:
Does this mean that `global_atomic_add v[1:2], v2, off` will work as before, but `global_atomic_add v[1:2], v2, null` won't be accepted anymore? This doesn't sound ideal if 'null' and 'off' are meant to be aliases (SP3 docs for GFX10 even say that 'off' is a legacy syntax and is translated to 'null').

jwanggit86 commented:
> Does this mean that `global_atomic_add v[1:2], v2, off` will work as before, but `global_atomic_add v[1:2], v2, null` won't be accepted anymore? This doesn't sound ideal if 'null' and 'off' are meant to be aliases (SP3 docs for GFX10 even say that 'off' is a legacy syntax and is translated to 'null').
The instruction `global_atomic_add v[1:2], v2, null` is not valid even without this patch. So already in this case `null` is not treated as an alias for `off`. The problem this patch is address is that the following 2 instructions have the same encoding:
```
global_atomic_add v[1:2], v2, off       ; encoding: [0x00,0x80,0xc8,0xdc,0x01,0x02,0x7d,0x00]
global_atomic_add v1, v2, null          ; encoding: [0x00,0x80,0xc8,0xdc,0x01,0x02,0x7d,0x00]
```

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIRegisterInfo.td:
```diff
@@ -845,6 +845,14 @@ def TTMP_64 : SIRegisterClass<"AMDGPU", [v2i32, i64, f64, v4i16, v4f16, v4bf16],
+
 def SReg_64_XEXEC : SIRegisterClass<"AMDGPU", [v2i32, i64, v2f32, f64, i1, v4i16, v4f16, v4bf16], 32,
   (add SGPR_64, VCC, FLAT_SCR, XNACK_MASK, SGPR_NULL64, SRC_SHARED_BASE,
        SRC_SHARED_LIMIT, SRC_PRIVATE_BASE, SRC_PRIVATE_LIMIT, TTMP_64, TBA, TMA)> {
```
Good point!

jwanggit86 commented on llvm/test/MC/AMDGPU/flat-global.s:
```diff
@@ -212,8 +212,8 @@ global_store_dword v3, v1, s[2:3] offset:-8
 global_store_dword v3, v1, exec
-// GFX10: encoding: [0x00,0x80,0x70,0xdc,0x03,0x01,0x7e,0x00]
-// GFX9: global_store_dword v3, v1, exec ; encoding: [0x00,0x80,0x70,0xdc,0x03,0x01,0x7e,0x00]
+// GFX10-ERR: :[[@LINE-1]]:{{[0-9]+}}: error: invalid operand for instruction
```
Will do.

jwanggit86 committed:
This commit: (1) define SReg_64_XEXEC in terms of SReg_64_XEXEC_XNULL
(2) fix tests affected by the new reg class. Todo: in final commit
msg, mention EXEC as well.

jayfoad commented:
> > Does this mean that `global_atomic_add v[1:2], v2, off` will work as before, but `global_atomic_add v[1:2], v2, null` won't be accepted anymore? This doesn't sound ideal if 'null' and 'off' are meant to be aliases (SP3 docs for GFX10 even say that 'off' is a legacy syntax and is translated to 'null').
> 
> The instruction `global_atomic_add v[1:2], v2, null` is not valid even without this patch. So already in this case `null` is not treated as an alias for `off`.

I agree that this patch does not break anything that currently works correctly. I also agree with Diana that it would be even better if `null` was treated the same as `off`, so that the assembler accepted `global_atomic_add v[1:2], v2, null`.

rovka commented:
> The instruction `global_atomic_add v[1:2], v2, null` is not valid even without this patch. 

Ok, in that case I'm happy with this patch after you address Jay's concern. It would be nice to follow up with another patch making `null` work wherever `off` works, if you have the time.

jwanggit86 commented:
> > The instruction `global_atomic_add v[1:2], v2, null` is not valid even without this patch.
> 
> Ok, in that case I'm happy with this patch after you address Jay's concern. It would be nice to follow up with another patch making `null` work wherever `off` works, if you have the time.

Yes, I can do that.

jwanggit86 committed:
Use declared reg class for SADDR instead of hard-code.

jwanggit86 commented:
> > > Does this mean that `global_atomic_add v[1:2], v2, off` will work as before, but `global_atomic_add v[1:2], v2, null` won't be accepted anymore? This doesn't sound ideal if 'null' and 'off' are meant to be aliases (SP3 docs for GFX10 even say that 'off' is a legacy syntax and is translated to 'null').
> > 
> > 
> > The instruction `global_atomic_add v[1:2], v2, null` is not valid even without this patch. So already in this case `null` is not treated as an alias for `off`.
> 
> I agree that this patch does not break anything that currently works correctly. I also agree with Diana that it would be even better if `null` was treated the same as `off`, so that the assembler accepted `global_atomic_add v[1:2], v2, null`.

Will address `global_atomic_add v[1:2], v2, null` in another PR.

jwanggit86 committed:
Resolve merge conflicts.

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -6215,6 +6215,8 @@ void SIInstrInfo::legalizeOperandsFLAT(MachineRegisterInfo &MRI,
 
   Register ToSGPR = readlaneVGPRToSGPR(SAddr->getReg(), MI, MRI);
+  if (MRI.getRegClass(ToSGPR) == &AMDGPU::SReg_64RegClass)
+    MRI.setRegClass(ToSGPR, &AMDGPU::SReg_64_XEXEC_XNULLRegClass);
```
Done. Changed to declared reg class.

jwanggit86 commented:
@jayfoad @arsenm ping.

jwanggit86 commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -6205,6 +6205,12 @@ void SIInstrInfo::legalizeOperandsFLAT(MachineRegisterInfo &MRI,
+      getRegClass(MI.getDesc(), SAddr->getOperandNo(),
+                  MRI.getTargetRegisterInfo(), *MI.getParent()->getParent());
+
+  MRI.setRegClass(ToSGPR, DeclaredRC);
```
For the following mir code:
```
  bb.0:
    liveins: $vgpr0_vgpr1
      %0:sreg_64 = COPY $vgpr0_vgpr1

  bb.1:
      %1:sreg_64_xexec_xnull = PHI %0, %bb.0, %2, %bb.1
      %3:vgpr_32 = V_MOV_B32_e32 1, implicit $exec
      %4:vgpr_32 = GLOBAL_LOAD_DWORD_SADDR %1, %3, 0, 0, implicit $exec
      %2:sreg_64 = S_AND_B64 %1, 1, implicit-def $scc
      S_CMP_LG_U64 %2, 0, implicit-def $scc
      S_CBRANCH_SCC1 %bb.1, implicit $scc

  bb.2:
      S_ENDPGM 0
```
In the SIFixSGPRCopies pass, `legalizeOperandsFLAT()` is called for the `GLOBAL_LOAD_DWORD_SADDR` instruction in the following call trace:
 ```
#0  llvm::SIRegisterInfo::getSGPRClassForBitWidth (BitWidth=64)
#1  0x0000555556ab1032 in llvm::SIRegisterInfo::getEquivalentSGPRClass (this=0x7ffff7e31380,
    VRC=0x55555f9eaaa0 <llvm::AMDGPU::VReg_64RegClass>)
#2  0x000055555697163b in llvm::SIInstrInfo::readlaneVGPRToSGPR (this=0x7ffff7e31330, SrcReg=..., UseMI=..., MRI=...)
#3  0x00005555569721c3 in llvm::SIInstrInfo::legalizeOperandsFLAT (this=0x7ffff7e31330, MRI=..., MI=...)
#4  0x0000555556974904 in llvm::SIInstrInfo::legalizeOperands (this=0x7ffff7e31330, MI=..., MDT=0x55555fc25558)
#5  0x0000555556978740 in llvm::SIInstrInfo::moveToVALUImpl (this=0x7ffff7e31330, Worklist=..., MDT=0x55555fc25558, Inst=...)
#6  0x0000555556976bbf in llvm::SIInstrInfo::moveToVALU (this=0x7ffff7e31330, Worklist=..., MDT=0x55555fc25558)
#7  0x0000555556901209 in (anonymous namespace)::SIFixSGPRCopies::lowerSpecialCase (this=0x55555fc24e80, MI=..., I=...)
```
Here in `readlaneVGPRToSGPR ()` it tries to get the SGPR equivalence of the src reg class, `VReg_64RegClass`. This leads to the call of `getSGPRClassForBitWidth()`, which, for 64b always returns `SReg_64RegClass`, which is not what we want here.
```cpp
const TargetRegisterClass *
SIRegisterInfo::getSGPRClassForBitWidth(unsigned BitWidth) {
  if (BitWidth == 16)
    return &AMDGPU::SGPR_LO16RegClass;
  if (BitWidth == 32)
    return &AMDGPU::SReg_32RegClass;
  if (BitWidth == 64)
    return &AMDGPU::SReg_64RegClass;
...
```
That's why in `legalizeOperandsFLAT()` we have to override it with the desired reg class for SAddr.

arsenm replied:
This is the wrong way to fix this. readlaneVGPRToSGPR can already check the operand class directly, and then apply getCommonSubClass to the most general result getSGPRClassForBitWidth returned. You shouldn't let it create the wrong register class and then redo it after

jwanggit86 replied:
> This is the wrong way to fix this. readlaneVGPRToSGPR can already check the operand class directly, and then apply getCommonSubClass to the most general result getSGPRClassForBitWidth returned. You shouldn't let it create the wrong register class and then redo it after

To use `getCommonSubClass()` as you suggested, I need to add an extra parameter to `readlaneVGPRToSGPR()` which holds the RC I want with a default of nullptr, as follows:
```diff
 Register SIInstrInfo::readlaneVGPRToSGPR(Register SrcReg, MachineInstr &UseMI,
-                                         MachineRegisterInfo &MRI) const {
+             MachineRegisterInfo &MRI, const TargetRegisterClass *DstRC/*=nullptr*/) const {
   const TargetRegisterClass *VRC = MRI.getRegClass(SrcReg);
   const TargetRegisterClass *SRC = RI.getEquivalentSGPRClass(VRC);
+  if (DstRC) {
+    SRC = ST.getRegisterInfo()->getCommonSubClass(SRC, DstRC);
+  }
   Register DstReg = MRI.createVirtualRegister(SRC);
```
Pls let me know if this is what you had in mind.

arsenm replied:
Something like this, (but you can just use RI directly as a member of SIInstrInfo)

Also it's easier to review changes when you post them as real revisions

jwanggit86 committed:
Pass desired reg class to readlaneVGPRToSGPR() so that it can get
the correct SGPR class.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -6048,10 +6048,14 @@ void SIInstrInfo::legalizeOperandsVOP3(MachineRegisterInfo &MRI,
   const TargetRegisterClass *VRC = MRI.getRegClass(SrcReg);
   const TargetRegisterClass *SRC = RI.getEquivalentSGPRClass(VRC);
+  if (DstRC)
+    SRC = ST.getRegisterInfo()->getCommonSubClass(SRC, DstRC);
```
```suggestion
    SRC = RI.getCommonSubClass(SRC, DstRC);
```

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.h:
```diff
@@ -1214,7 +1214,8 @@ class SIInstrInfo final : public AMDGPUGenInstrInfo {
   /// threads in the wave.
   /// \returns The SGPR register that \p SrcReg was copied to.
```
Comment the parameter

jwanggit86 replied:
Done.

jwanggit86 committed:
Minor code changes.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -6244,7 +6248,11 @@ void SIInstrInfo::legalizeOperandsFLAT(MachineRegisterInfo &MRI,
-  Register ToSGPR = readlaneVGPRToSGPR(SAddr->getReg(), MI, MRI);
+  const TargetRegisterClass *DeclaredRC =
+      getRegClass(MI.getDesc(), SAddr->getOperandNo(),
+                  MRI.getTargetRegisterInfo(), *MI.getParent()->getParent());
```
```suggestion
                  &RI, *MI.getParent()->getParent());
```

jwanggit86 replied:
Done

jwanggit86 committed:
Minor code change.

jwanggit86 committed:
A test that works fine locally but fails llvm build.
